core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala

```diff
@@ -463,7 +430,24 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
       _ = logger.info(
         s"Created Azure storage container in WSM [workspaceId = ${workspaceId}, containerId = ${containerResult.getResourceId}]"
       )
-    } yield savedWorkspace
+    } yield savedWorkspace).recoverWith { case e @ (_: ApiException | _: WorkspaceManagerCreationFailureException) =>
```

Comment:
Is there any chance WSM gives back a 409 in case there's a duplicate? Thinking through the retry case here.

Reply:
Good catch. 

If the workspace id is already in use when WSM first receives the request, it will [return a 409](https://github.com/DataBiosphere/terra-workspace-manager/blob/f607490bea34363768e5d9cb4c4ba5d379ddd4fa/service/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java#L143). If one of the later steps in the workspace creation flight fails due to a duplicate workspace, WSM will return a 400 instead 🙃 (see [`WorkspaceService`](https://github.com/DataBiosphere/terra-workspace-manager/blob/f607490bea34363768e5d9cb4c4ba5d379ddd4fa/service/src/main/java/bio/terra/workspace/service/workspace/WorkspaceService.java#L113) and [`CreateWorkspaceStep`](https://github.com/DataBiosphere/terra-workspace-manager/blob/f607490bea34363768e5d9cb4c4ba5d379ddd4fa/service/src/main/java/bio/terra/workspace/service/workspace/flight/CreateWorkspaceStep.java#L39) which both throw [`DuplicateWorkspaceException`](https://github.com/DataBiosphere/terra-workspace-manager/blob/f607490bea34363768e5d9cb4c4ba5d379ddd4fa/service/src/main/java/bio/terra/workspace/service/workspace/exceptions/DuplicateWorkspaceException.java#L6) which extends `BadRequest`, not `Conflict`).

WSM only checks the workspace id for uniqueness. Given the low likelihood of a UUID collision, I question whether it's worth accounting for it but 🤷